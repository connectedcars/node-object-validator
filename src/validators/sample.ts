import {
  isValidator,
  ValidateOptions,
  ValidatorBase,
  ValidatorBaseOptions,
  ValidatorExportOptions,
  ValidatorOptions
} from '../common'
import { ValidationFailure } from '../errors'
import { RequiredArray } from './array'
import { RequiredBoolean } from './boolean'
import { RequiredDate } from './date'
import { isDateTime, RequiredDateTime } from './datetime'
import { RequiredFloat } from './float'
import { RequiredInteger } from './integer'
import { RequiredNull } from './null'
import { ObjectSchema, RequiredObject } from './object'
import { RequiredRegexMatch } from './regex-match'
import { RequiredString } from './string'
import { OptionalUnknown, RequiredUnknown } from './unknown'

export type Sample =
  | null
  | boolean
  | number
  | string
  | Sample[]
  | Date
  | RegExp
  | ValidatorBase
  | undefined
  | { [prop: string]: Sample | undefined }

export function sampleToValidator(sample: Sample, options?: ValidatorBaseOptions): ValidatorBase {
  if (sample === null) {
    return new RequiredNull(options)
  }
  if (isValidator(sample)) {
    return sample
  }

  switch (typeof sample) {
    case 'undefined': {
      return new OptionalUnknown(options)
    }
    case 'object': {
      if (Array.isArray(sample)) {
        if (sample.length === 0) {
          return new RequiredArray(new RequiredUnknown(options))
        } else {
          return new RequiredArray(
            isValidator(sample[0]) ? sample[0] : sampleToValidator(sample[0]),
            0,
            Number.MAX_SAFE_INTEGER,
            options
          )
        }
      } else if (sample instanceof Date) {
        return new RequiredDate(options)
      } else if (sample instanceof RegExp) {
        return new RequiredRegexMatch(sample, options)
      } else {
        const objectSchema: ObjectSchema = {}
        for (const key of Object.keys(sample)) {
          const prop = sample[key]
          objectSchema[key] = sampleToValidator(prop)
        }
        return new RequiredObject(objectSchema, options)
      }
    }
    case 'number': {
      if (Number.isInteger(sample)) {
        return new RequiredInteger(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, options)
      }
      return new RequiredFloat(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, options)
    }
    case 'string': {
      if (isDateTime(sample)) {
        return new RequiredDateTime(options)
      } else {
        return new RequiredString(0, Number.MAX_SAFE_INTEGER, options)
      }
    }
    case 'boolean': {
      return new RequiredBoolean(options)
    }
  }
}

export function isSample<T extends Sample>(sample: T, value: unknown, context?: string): value is SampleWrap<T> {
  const errors = validateSample(sample, value, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateSample(
  sample: Sample,
  value: unknown,
  context?: string,
  options?: ValidateOptions
): ValidationFailure[] {
  const validator = sampleToValidator(sample, options)
  return validator.validate(value, context, options)
}

// prettier-ignore
type SampleWrap<O> =
  O extends ValidatorBase ? O['tsType'] :
  O extends number ? number  :
  O extends string ? string  :
  O extends boolean ? boolean : O

export abstract class SampleValidator<T extends Sample = never, O = never> extends ValidatorBase<SampleWrap<T> | O> {
  public schema: T
  private validator: ValidatorBase

  public constructor(sample: T, options?: ValidatorBaseOptions) {
    super(options)
    this.schema = sample
    this.validator = sampleToValidator(this.schema, options)
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    return this.validator.validate(value, context, { earlyFail: this.earlyFail, ...options })
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = this.validator.toString(options)

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        throw new Error('Rust not supported yet')
      }
      default: {
        throw new Error(`Language: '{}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredSample<T extends Sample> extends SampleValidator<T> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options })
  }
}

export class OptionalSample<T extends Sample> extends SampleValidator<T, undefined> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false })
  }
}

export class NullableSample<T extends Sample> extends SampleValidator<T, null> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, nullable: true })
  }
}

export class OptionalNullableSample<T extends Sample> extends SampleValidator<T, null | undefined> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false, nullable: true })
  }
}
