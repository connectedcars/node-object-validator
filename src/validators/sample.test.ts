import {
  DoesNotMatchRegexFail,
  NotDateFail,
  NotFloatFail,
  NotIntegerFail,
  NotStringFail,
  RequiredFail
} from '../errors'
import { SampleValidator, validateSample } from './sample'

describe('Sample', () => {
  describe('validateSample', () => {
    it('should validate array sample', () => {
      expect(validateSample([], ['value1', 1])).toEqual([])
      expect(validateSample([''], ['value1', 'value2'])).toEqual([])
      expect(validateSample(['status'], ['status', 'status'])).toEqual([])
      expect(validateSample([1], [1, 2])).toEqual([])
      expect(validateSample([1.1], [1.1, 2.1])).toEqual([])
      expect(validateSample([/^status[0-9]$/], ['status1', 'status2'])).toEqual([])
      expect(validateSample([new Date()], [new Date(), new Date()])).toEqual([])
      expect(
        validateSample(
          [
            {
              type: 'message'
            }
          ],
          [{ type: 'message' }]
        )
      ).toEqual([])
      expect(validateSample(['2018-08-06T13:37:00.000Z'], ['2018-08-06T13:37:00.000Z'])).toEqual([])
    })

    it('should validate Date sample', () => {
      expect(validateSample(new Date(), new Date())).toEqual([])
      expect(validateSample(new Date(), '')).toEqual([new NotDateFail('Must be a Date object')])
    })

    it('should validate iso date time sample', () => {
      expect(validateSample('2018-08-06T13:37:00.000Z', '2018-08-06T13:37:00.000Z')).toEqual([])
      expect(validateSample('2018-08-06T13:37:00.000Z', '')).toEqual([])
    })

    it('should validate float sample', () => {
      expect(validateSample(1.1, 2.1)).toEqual([])
      expect(validateSample(1.1, '')).toEqual([new NotFloatFail('Must be a float (received "")')])
    })

    it('should validate integer sample', () => {
      expect(validateSample(1, 2)).toEqual([])
      expect(validateSample(1, '')).toEqual([new NotIntegerFail('Must be an integer (received "")')])
    })

    it('should validate object sample', () => {
      const schema = {
        type: 'message'
      }
      expect(validateSample(schema, { type: 'message' })).toEqual([])
      expect(validateSample(schema, { value: 'message' })).toEqual([new RequiredFail('Is required', { key: 'type' })])
    })

    it('should validate regex sample', () => {
      expect(validateSample(/^status$/, 'status')).toEqual([])
      expect(validateSample(/^status$/, 'something else')).toEqual([
        new DoesNotMatchRegexFail(`Did not match '/^status$/' (received "something else")`)
      ])
    })

    it('should validate string sample', () => {
      expect(validateSample('', '')).toEqual([])
      expect(validateSample('', 2)).toEqual([new NotStringFail('Must be a string (received "2")')])
    })
  })
})

describe.each([false, true])('Sample (optimize: %s)', optimize => {
  describe('SampleValidator', () => {
    it('should validate an array', () => {
      const arrayValidator = new SampleValidator([1], { optimize })
      const errors = arrayValidator.validate([1, 2, 4, 5])
      expect(errors).toEqual([])
    })

    it('should validate a complex object', () => {
      const sample = {
        type: 'gps_odometer_km',
        unitId: '1234567',
        recordedAt: '2018-08-06T13:37:00Z',
        tripId: 1337,
        value: 500,
        position: {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18,
          extra: {
            tag: 'test',
            tagversion: 32,
            tagDepth: 3.1416
          }
        },
        positions: [
          {
            latitude: 55.332131,
            longitude: 12.54454,
            accuracy: 18
          }
        ]
      }
      const sampleValidator = new SampleValidator(sample, { optimize })
      const errors = sampleValidator.validate(sample)
      expect(errors).toEqual([])
    })
  })
})
