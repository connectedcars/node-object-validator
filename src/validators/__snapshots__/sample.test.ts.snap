// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sample RequiredSample should return a function body 1`] = `
Array [
  Object {
    "RequiredFail": [Function],
  },
  Array [
    "const validator1 = validator1",
  ],
  Array [
    "if (value1 !== undefined) {",
    "  errors.push(...validator1.validateValue(value1, context))",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
